### PR TITLE
ARROW-17518: [CI][Doc][Python] Update glob to detect arrow development version from git

### PR DESCRIPTION
**Pull request by raulcd**
Reproduced locally:
```
$ git describe --dirty --tags --long --match "apache-arrow-[0-9].*"
apache-arrow-9.0.0.dev-641-g0d5bb92-dirty
$ git describe --dirty --tags --long --match "apache-arrow-[0-9]*.*"
apache-arrow-10.0.0.dev-114-g0d5bb92-dirty
```

### CODE REVIEW

File: dev/archery/archery/crossbow/core.py [hunk 1]
```diff
@@ -696,7 +696,7 @@ def get_version(root, **kwargs):
 
     # query the calculated version based on the git tags
     kwargs['describe_command'] = (
-        'git describe --dirty --tags --long --match "apache-arrow-[0-9].*"'
+        'git describe --dirty --tags --long --match "apache-arrow-[0-9]*.*"'
```
Comment (pitrou) [inline dev/archery/archery/crossbow/core.py:699]: I would mandate at least one digit
```suggestion
        'git describe --dirty --tags --long --match "apache-arrow-[0-9]+.*"'
```

File: python/pyarrow/__init__.py [hunk 2]
```diff
@@ -51,7 +51,7 @@ def parse_git(root, **kwargs):
             """
             from setuptools_scm.git import parse
             kwargs['describe_command'] = \
-                "git describe --dirty --tags --long --match 'apache-arrow-[0-9].*'"
+                "git describe --dirty --tags --long --match 'apache-arrow-[0-9]*.*'"
```
Comment (pitrou) [inline python/pyarrow/__init__.py:54]: Same here
```suggestion
                "git describe --dirty --tags --long --match 'apache-arrow-[0-9]+.*'"
```

File: python/setup.py [hunk 3]
```diff
@@ -586,7 +586,7 @@ def parse_git(root, **kwargs):
     """
     from setuptools_scm.git import parse
     kwargs['describe_command'] =\
-        'git describe --dirty --tags --long --match "apache-arrow-[0-9].*"'
+        'git describe --dirty --tags --long --match "apache-arrow-[0-9]*.*"'
```
Comment (pitrou) [inline python/setup.py:589]: Wow, how many instances of this do we have? :-)
```suggestion
        'git describe --dirty --tags --long --match "apache-arrow-[0-9]+.*"'
```

**Review by pitrou — COMMENTED** [review]
Thanks for finding the issue @raulcd !
Can you mark this PR ready?

**Comment by pitrou** [comment]
Also, can you please rebase to get a passing AppVeyor build?

File: dev/archery/archery/crossbow/core.py (hunk 1 above)
Comment (raulcd) [reply dev/archery/archery/crossbow/core.py:699]: I don't think that works as is git uses glob not regex. I did try that first :) :
```
$ git describe --dirty --tags --long --match "apache-arrow-[0-9]+.*"
fatal: No names found, cannot describe anything.
```

Comment (pitrou) [reply dev/archery/archery/crossbow/core.py:699]: Ahah, I see. Git being git, as usual...

**Comment by raulcd** [comment]
As it can be seen on the builds: https://github.com/apache/arrow/runs/8010907934?check_suite_focus=true#step:6:1322
the new version is selected correctly:
`ARROW_GIT_DESCRIPTION=apache-arrow-10.0.0.dev-120-g4e1922569`
or
`pyarrow==10.0.0.dev120+g4e1922569`
which was not correct previously.

**Review by pitrou — APPROVED** [review]
Thanks for diagnosing and fixing this @raulcd :-)

**Pull request merged**